kernel/src/apic.c

```diff
@@ -0,0 +1,296 @@
+#include "apic.h"
+
+#include "acpi.h"
+#include "port_io.h"
+#include "memory/paging.h"
+#include "kassert.h"
+
+#define MAX_IOAPIC_COUNT 4
+#define MAX_LAPIC_COUNT 16
+
+#define LOCAL_APIC_ENTRY 0
+#define IO_APIC_ENTRY 1
+#define INTERRUPT_SOURCE_OVERRIDE_ENTRY 2
+#define NON_MASKABLE_INTERRUPTS_ENTRY 4
+#define LOCAL_APIC_ADDRESS_OVERRIDE_ENTRY 5
+
+#define APIC_BASE_MSR 0x1b
+#define APIC_BASE_MSR_ENABLE 0x800
+
+// Gets lower register offset for irq redtable entry
+#define REDTBL_OFFSET(irq) (2 * irq + 0x10)
```

Comment:
You should enclose the arguments in both of these macros with parentheses.
Example:
```c
#define REDTBL_OFFSET(irq) (2 * (irq) + 0x10)
```